src/proxy/http/metrics/handle_time.rs

```diff
@@ -0,0 +1,252 @@
+use metrics::{latency, FmtLabels, FmtMetric, Histogram};
+use proxy::http::insert;
+use std::{
+    fmt,
+    sync::{
+        atomic::{self, AtomicUsize, Ordering},
+        Arc, Mutex, RwLock,
+    },
+    time::Instant,
+};
+
+/// A single handle time histogram.
+///
+/// Higher-level code will use this to represent a single set of labels for
+/// handle-time metrics.
+#[derive(Debug, Clone)]
+pub struct Scope(Arc<Shared>);
+
+/// A layer that inserts a `Tracker` into each request passing through it.
+#[derive(Debug, Clone)]
+pub struct InsertTracker(Arc<Shared>);
+
+/// A request extension that, when dropped, records the time elapsed since it
+/// was created.
+#[derive(Debug)]
+pub struct Tracker {
+    shared: Arc<Shared>,
+    idx: usize,
+    t0: Instant,
+}
+
+#[derive(Debug)]
+struct Shared {
+    // NOTE: this is inside a `Mutex` since recording a latency requires a mutable
+    // reference to the histogram. In the future, we could consider making the
+    // histogram counters `AtomicU64, so that the histogram could be updated
+    // with an immutable reference. Then, the mutex could be removed.
+    histogram: Mutex<Histogram<latency::Us>>,
+    /// Stores the state of currently active `Tracker`s.
+    recorders: RwLock<Vec<Recorder>>,
+    /// The index of the most recently finished recorder.
+    ///
+    /// When a new recorder is needed, the recorder at this index will be used,
+    /// and that recorder's next index will be set as the head. of the free
+    /// list.
+    ///
+    /// When an active recorder completes, this will be set to its index, and
+    /// the previous value will become the freed recorder's next pointer.
+    idle_head: AtomicUsize,
+}
+
+#[derive(Debug)]
+struct Recorder {
+    /// The number of currently active `Tracker`s for this recorder.
+    ///
+    /// When a request is initially received, there will be one `Tracker` in its
+    /// `Extensions` map. If the request is cloned for retries, the `Tracker`
+    /// will be cloned, incrementing this count. Dropping a `Tracker` decrements
+    /// this count, and when it reaches 0, the handle time is recorded.
+    active_trackers: AtomicUsize,
+    /// Index of the next free recorder.
+    next_idle: AtomicUsize,
+}
+
+impl insert::Lazy<Tracker> for InsertTracker {
+    fn value(&self) -> Tracker {
+        self.0.clone().tracker()
+    }
+}
+
+// ===== impl Scope =====
+
+impl Scope {
+    pub fn new() -> Self {
```

Comment:
nit: i tend to prefer the module fn pattern of pub fn handle_time() -> (Scope, InsertLayer) -- using the scope to produce layers is fine, but also doesn't really seem necessary (since the layer should be clone)...